### ISSUE TITLE
Add support for assigning IPv6 prefixes to EC2 instances

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#19251]
This is the feature request for the IPv6 equivalent to #16987, which should follow a similar implementation to https://github.com/cilium/cilium/pull/18463.

**Comment by bertjwregeer** [comment]
It's not stale, its still something that should get added :-)

**Comment by danehans** [comment]
> should follow a similar implementation to https://github.com/cilium/cilium/pull/18463

@joestringer ENIs are assigned /80 prefixes for IPv6 prefix delegation. The prefix length is not configurable and due to the number of addresses supported by a /80 prefix, surfacing available ENI IPv6 addresses through `ciliumnode.spec.ipam.pool` can be a UX and scalability issue. LMK if you have an issue with creating a restriction to the number of IPv6 addresses that can be allocated from an IPv6 prefix so the implementation can be similar to #18463.

**Comment by joestringer** [comment]
I think you were able to get a response from the community meeting on this? I was traveling last week so I'm a bit behind on all of my notifications. Originally when I filed this issue, I assumed that whatever design we have for IPv4 is reasonable to also apply for IPv6 and so that was the background to "following the similar implementation". AFAIK though the scalability realities of IPv6 mean that it's worth revisiting some of those baseline details, is that accurate? I don't have specific knowledge about the implementation/design of this, but I'm sure that @cilium/sig-ipam would be happy to discuss further.

**Comment by danehans** [comment]
@joestringer thanks for circling back on this issue. The following feedback on this topic was captured during last week's community meeting:

> @hemanthmalla: Do we really need to enumerate all IPs listed on the CRD?
> @EItanya: This is a divergence from how IPAM is currently managed.

If I correctly understand @hemanthmalla's point, I do question the need to surface available IPs via `ciliumnodes.spec.ipam.pool`. However, IPv6 prefix delegation should provide a consistent UX with its IPv4 equivalent. We could potentially use `ciliumnodes.spec.ipam.pools` to surface delegated IPv4/IPv6 ENI prefixes instead of `ciliumnodes.spec.ipam.pool`, but to maintain a consistent UX this change should start with IPv4 prefix delegation before adding IPv6 prefix delegation. However, I don't believe this change is possible since the CiliumNode API type is GA'd. Due to these points, I think it's reasonable to limit the number of IPs that are available from delegated /80 ENI IPv6 prefixes and follow the existing IPv4 implementation. #30684 is the first step towards this approach. @cilium/sig-ipam thoughts?

**Comment by joestringer** [comment]
AFAIK `ciliumnodes.spec.ipam.pool` usage is part of the protocol between the cilium-agents and cilium-operator for scalable use of cloud IPAM APIs, specifically with particular (cloud) IPAM modes. Related, ENI IPAM docs: https://docs.cilium.io/en/stable/network/concepts/ipam/eni/ . I'm not sure how this intersects with more recent multi-pool designs though.

> However, I don't believe this change is possible since the CiliumNode API type is GA'd.

I wouldn't consider CiliumNode "done" but yes if we extend it then we need to be careful about not breaking existing things. I believe we can update the CRD and extend it to include new fields, assuming we don't break existing functionality and assuming that there's a reasonable design for the extension.

**Comment by EItanya** [comment]
> AFAIK ciliumnodes.spec.ipam.pool usage is part of the protocol between the cilium-agents and cilium-operator for scalable use of cloud IPAM APIs, specifically with particular (cloud) IPAM modes.

This is correct, the problem is that by its nature IPV6 supports many many more IPs than IPV4, so I was worried that enumerating all of the used IPs on the Node Status would become unwieldy. However, the more I think about it I think we'll be ok. First of all, the list is per node, and I doubt anyone would try to run TOO many pods on a single node, willing to be proven wrong on that point though. Second, we could impose maximum number of allocatable IPs which is user configurable if we are worried about people scheduling too many pods on single nodes and allowing the status to get larger than `etcd` size limits

**Comment by marseel** [comment]
> First of all, the list is per node, and I doubt anyone would try to run TOO many pods on a single node, willing to be proven wrong on that point though. Second, we could impose maximum number of allocatable IPs which is user configurable if we are worried about people scheduling too many pods on single nodes and allowing the status to get larger than etcd size limits

I would add my 2 cents, that Etcd size limit is usually not the biggest concern. The main issue with having IPs in CiliumNode is that all agents are watching all CiliumNodes so whenever we are updating CiliumNode with new IPs, all agents receive an update and process it (and then just ignore the event). Unfortunately, I don't think there is a simple solution to that problem except for moving IP allocations to separate CRD (so each agent watches only changes related to their own node) - still the same problem already exists for IPv4 -  we don't need to solve it here, but just something to keep in mind. 

I would maybe consider just assigning a fixed amount of IPv6, like 250 and **not** reporting back which IPs are used - so we don't generate a higher churn of updates on CiliumNode and from the scalability point of view, IPv6 would be more scalable than IPv4 current implementation :)

**Comment by danehans** [comment]
@marseel thanks for sharing your insights. I'll look into __not__ reporting used IPs in the CiliumNode status. My only concern is that the UX will differ from IPv4 but maybe that's a limitation that users live with until the CRD-based approach can be implemented.

**Comment by joestringer** [comment]
How does updating the status impact updates for the actual resource? Surely other nodes don't need to be notified if they are not monitoring for the status? I see the benefit for scalability but also the potential pitfalls for debuggability/observability, so if there's a middle ground where we can get a bit of both then that would be a "nice to have".

**Comment by marseel** [comment]
>My only concern is that the UX will differ from IPv4 but maybe that's a limitation that users live with until the CRD-based approach can be implemented.

Definitely makes sense, I also don't like potential UX change here. Migration though might be significantly harder later on.

> How does updating the status impact updates for the actual resource? Surely other nodes don't need to be notified if they are not monitoring for the status?

Status is just part of the resource so if you are watching resource changes == you are watching status changes and if you are updating status == you are updating resource.

**Comment by HadrienPatte** [comment]
Hey, coming back to this issue and taking into account the feedback from previous implementation attempts as well as how the cilium project has evolved since then, I have a proposal for a potential plan to get IPv6 support for ENI IPAM mode with full support for prefix delegation. 

Previous attempts at bringing IPv6 support essentially ported the existing IPv4 prefix delegation logic to IPv6. But the main issue with that approach was that the current IPv4 prefix delegation implementation flattens IPs from each prefix in `CiliumNode.Spec.IPAM.Pool` and agents then handle those IPs through the CRD allocator just like regular secondary IPs. This works for IPv4 as there are only 16 IPs in a /28 delegated prefix, but flattening all IPs from a prefix like that is just not an option for IPv6 where delegated prefixes are /80 (2^48 IPs).

My proposal would be to essentially have the operator advertise delegated prefixes without flattening them, for example via the existing `CiliumNode.Spec.IPAM.Pools.Allocated` and have the agents assign IPs from those pools by reusing the existing multipool allocator logic instead of the CRD allocator that's currently used in ENI IPAM mode.

Now directly implementing ENI IPv6 support like that may not be desirable as that would put us in a place where ENI mode is implemented significantly differently between both IP families. That's why I'd suggest starting with reworking the existing IPv4 prefix delegation logic to start using a multipool allocator based implementation. After that's done, IPv6 support can then be added by porting over that updated IPv4 implementation.

As an illustration of what I'm proposing with IPv4, here's the current flow when a prefix gets assigned:
* operator sets `ENI.Prefixes = ["10.0.0.0/28"]`
* operator expands prefix in `Spec.IPAM.Pool = {"10.0.0.0": …, "10.0.0.1": …, … ×16}`
* agent allocates individual IPs with `Status.IPAM.Used` updated per pod
* operator scans all 16 IPs to detect idle prefix and call `UnassignENIPrefixes`

And here's what the new flow would look like:
* operator sets `ENI.Prefixes = ["10.0.0.0/28"]`
* operator sets `Spec.IPAM.Pools.Allocated = [{pool:"eni", cidrs:["10.0.0.0/28"]}]`
* agent's cidrPool allocates IPs locally, no per-IP K8s updates
* agent writes demand in `Spec.IPAM.Pools.Requested = [{pool:"eni", needed:{ipv4Addrs: N}}]`
* operator allocates more /28s if demand > current CIDR capacity
* agent removes idle /28 from Allocated and operator calls `UnassignENIPrefixes`

Happy to get any feedback on this proposal

**Comment by gandro** [comment]
> Happy to get any feedback on this proposal

This sounds great! Really happy to see this being revived. Two bits of (non-blocking) feedback:

First, I agree that with some of the more recent refactors of multi-pool, it probably should be possible to have the ENI agent allocator use bits of the multi-pool allocator under the hood. Overall, ENI still needs some special handling (like setting up egress IP rules and interface setup), but I assume that's part of your plan as well. One issue with Multi-Pool at the moment is that we don't allow the full /28 to be used, the allocator reserves the first and last IP address for special purposes. This is fixable, but might need some care. See also https://github.com/cilium/cilium/issues/28637#issuecomment-2337426529

Second, and this really more a of a wishlist item for the far future - I would love for users to be able to control what ENI subnet a particular pod is placed in. At the moment, this is only controllable on a per-node level (using subnet-tags-filter etc in the CiliumNode). By introducing the concept of pools to ENI mode, it would be nice if we made sure that maybe using custom pool names (rather than just the proposed "eni" name here) in the future would be possible.

**Comment by HadrienPatte** [comment]
> Overall, ENI still needs some special handling (like setting up egress IP rules and interface setup), but I assume that's part of your plan as well.

Yes, I identified this point, and we should be able to solve this by extracting [`configureENIDevices`](https://github.com/cilium/cilium/blob/4554ebf83dff2336c1fdb4f2d3132d15eb1486d4/pkg/ipam/crd_eni.go#L36) from [`nodeStore.updateLocalNodeResource`](https://github.com/cilium/cilium/blob/4554ebf83dff2336c1fdb4f2d3132d15eb1486d4/pkg/ipam/crd.go#L425) into a standalone `startENIDeviceWatcher` that runs its own `CiliumNode` informer. That would ensure ENI interfaces get configured regardless of which IPAM allocator path is used (CRD or multipool).

> One issue with Multi-Pool at the moment is that we don't allow the full /28 to be used, the allocator reserves the first and last IP address for special purposes.

That's a good point 👍 we do want those addresses to be assignable, especially for the IPv4 usecase where 2 addresses out of a /28 prefix represents 12.5% of addresses.

> it would be nice if we made sure that maybe using custom pool names (rather than just the proposed "eni" name here) in the future would be possible.

Agree, I still only have a high level overview of the proposed changes, but this definitely sounds like some new capabilities that a switch to a multipool based implementation could allow.

**Comment by HadrienPatte** [comment]
👋 I wrote up a more detailed design proposal as a CFP here: https://github.com/cilium/design-cfps/pull/87

**Comment by 41ks** [comment]
Following @HadrienPatte's work on migrating the ENI IPAM to multi-pool, I have started looking into actually implementing IPv6 support. The idea, as Hadrien explained, will be to just give agents a /80 and let them handle allocation. That way we don't run into any issues regarding how big that prefix is.

One point that I would like to raise is that it seems to me we can choose to only support prefix delegation mode for IPv6. Since subnets can have a minimum CIDR size of /64 (up to /50), that means each subnet has at least 65K /80 prefixes, which is more than enough for a Kubernetes cluster. I don't see any use case where it would be useful to add a single IPv6 address to an ENI.
We could also do without releasing the IPv6 prefix, since there will always be a maximum of one prefix per ENI and the subnet doesn't risk running out of prefixes. By not having to implement those mechanisms for IPv6, this should simplify ENI IPv6 support.

**Comment by 41ks** [comment]
I opened a PR #46619 to implement the operator side of the IPv6 prefix IPAM. Feel free to take a look !

The agent side needs further work on my end but I do have a working prototype which looks quite promising.

**Comment by HadrienPatte** [comment]
Closing this issue as implemented with #46619 and #46756. Will keep #18405 open to track followup improvements while IPv6 on ENI mode is in beta.
